src/linpde_gp/linops/_block.py

```diff
@@ -1,94 +1,201 @@
-                np.promote_types(self._C.dtype, self._D.dtype),
-            ),
-            matmul=self._matmul,
-            rmatmul=self._rmatmul,
-            apply=self._apply,
-            todense=lambda: np.block(
-                [
-                    [self._A.todense(cache=False), self._B.todense(cache=False)],
-                    [self._C.todense(cache=False), self._D.todense(cache=False)],
-                ]
-            ),
-            transpose=lambda: BlockMatrix(
-                A=self._A.T, B=self._C.T, C=self._B.T, D=self._D.T
-            ),
-            inverse=lambda: BlockInverse(self),
-            # det= TODO
-            trace=lambda: self._A.trace + self._D.trace,
+        dtype = np.promote_types(
+            self._A.dtype if B is None else np.promote_types(self._A.dtype, B.dtype),
+            self._D.dtype if C is None else np.promote_types(self._D.dtype, C.dtype),
         )
 
-    def _matmul(self, x):
-        x0, x1 = self._split_input(x, axis=-2)
-
-        return np.concatenate(
-            (
-                self._A @ x0 + self._B @ x1,
-                self._C @ x0 + self._D @ x1,
-            ),
-            axis=-2,
+        super().__init__(
+            shape=(A.shape[0] + D.shape[0], A.shape[1] + D.shape[1]), dtype=dtype
         )
 
-    def _rmatmul(self, x):
-        x0, x1 = self._split_input(x, axis=-1)
+        self.is_block_diagonal = False
+        if B is None and C is None:
+            # Block diagonal
+            self._B = pn.linops.Zero((self._A.shape[0], self._D.shape[1]), dtype)
+            self._C = pn.linops.Zero((self._D.shape[0], self._A.shape[1]), dtype)
+            self.is_block_diagonal = True
+            if self._A.is_symmetric and self._D.is_symmetric:
+                self.is_symmetric = True
+        elif is_spd:
+            assert (B is None) ^ (C is None)  # Exactly one of B or C must be None
+            assert self._A.is_symmetric and self._D.is_symmetric
+            assert self._A.is_positive_definite
+            if C is None:
+                self._B = pn.linops.aslinop(B)
+                self._C = self._B.T
+            else:
+                self._C = pn.linops.aslinop(C)
+                self._B = self._C.T
+            self.is_symmetric = True
+            self.is_positive_definite = True
+        elif is_triangular:
+            assert (B is None) ^ (C is None)
+            if B is None:
+                assert self._A.is_lower_triangular and self._D.is_lower_triangular
+                self._C = pn.linops.aslinop(C)
+                self._B = pn.linops.Zero((self._A.shape[0], self._D.shape[1]), dtype)
+                self.is_lower_triangular = True
+            else:
+                assert self._A.is_upper_triangular and self._D.is_upper_triangular
+                self._B = pn.linops.aslinop(B)
+                self._C = pn.linops.Zero((self._D.shape[0], self._A.shape[1]), dtype)
+                self.is_upper_triangular = True
```

Comment:
We should be able to infer this from
```python
(self._A.is_lower_triangular and self._D.is_lower_triangular and B is None)
or (self._A.is_upper_triangular and self._D.is_upper_triangular and C is None)
```
Then we can get rid of the `is_triangular` argument